### ISSUE TITLE
关于找不到apollo.meta的问题分析和解决

### ISSUE THREAD

**Issue by slankka** [ctripcorp/apollo#3072]
我想提供一个这个问题真正的原因分析和解决方案
## 问题描述

```
Meta server address fallback to http://apollo.meta for env PRO, because it is not available in all MetaServerProviders

Could not complete get operation [Cause: apollo.meta]
```

## 问题分析
![apollo](https://user-images.githubusercontent.com/8142133/81635394-bcd4be00-9443-11ea-8b4a-2059b449cbe6.png)

Apollo-client和Apollo-core包内的 这两个SPI文件实现类不一致（目前中央仓库1.6.0版本就是），在打可执行jar包的时候，有可能会覆盖另一个文件。这两个实现所读取的meta方式不同：
 - 在apollo-client发布到中央仓库之前，是读取`apollo-env.properties` (或文件内同名环境变量)
 - 在apollo-client发布到中央仓库之后，是读取`apollo.meta`这个环境变量。

## 解决方案
如果项目是可执行jar，一般解决办法就是：
```
<plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-shade-plugin</artifactId>
        <version>2.4.1</version>
        <executions>
          <execution>
            <phase>package</phase>
            <goals>
              <goal>shade</goal>
            </goals>
            <configuration>
              <createDependencyReducedPom>false</createDependencyReducedPom>
              <transformers>
          .....
                <transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
                  <resource>META-INF/services/com.ctrip.framework.apollo.core.spi.MetaServerProvider</resource>
                </transformer>
              </transformers>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

使用shade插件将两个文件Append到一起。这样就能真正实现'fallback'。

## 总结
**SpringBoot可执行jar的项目，会因打包的插件不同，以及使用者定义meta方式不同而具有不确定性。**

SpringBoot使用的打包插件spring-boot-maven-plugin亦是基于maven-shade-plugin，大多数情况项目并不会报错，这里就不花时间分析SpringBoot是否针对重复的SPI文件做了Appending操作。

如果遇到找不到apollo-meta，报错的原因知道了，便不再困惑，参考上述内容即可对症解决。

*如果觉得这个分享不错，希望能写入WIKI*

**Comment by nobodyiam** [comment]
为什么会有这个问题，因为用户把apollo的jar包做了maven shade?

**Comment by slankka** [comment]
@nobodyiam 是的。我刚刚看了，spring-boot-maven-plugin配置了`org.apache.maven.plugins.shade.resource.ServicesResourceTransformer`，而且打包出来的jar都是在lib下面的，SPI能找到这两种实现，因此不会出问题。

有些用户可能用的不是spring-boot-maven-plugin，自己配置了maven-shade-plugin，尤其是一些用于非web场景的可执行jar，在实践中，确实容易出问题。

**Comment by lzhaoyang** [comment]
想知道gradle方式打包如何解决，目前gradle打包后，使用命令行参数传入-Denv=* -Dsit_meta=*** 出现issue描述的问题

**Comment by slankka** [comment]
@lzhaoyang [Gradle Shadow Plugin
](https://imperceptiblethoughts.com/shadow/configuration/merging/#merging-service-descriptor-files)

**Comment by skyyan2000** [comment]
遇到同样的问题 ，我再properties里边配置了apollo.meta，结果不生效 不知道大家有什么好的方式没

**Comment by yemo-memeda** [comment]
> 遇到同样的问题 ，我再properties里边配置了apollo.meta，结果不生效 不知道大家有什么好的方式没

我也是这样子，目前还没解决

**Comment by nobodyiam** [comment]
@skyyan2000 @yemo-memeda 

这个 issue 之前的问题是 maven shade 后导致文件被覆盖所致，你们也是同样的问题吗？
